crates/primitives/src/trie/hash_builder/proof_retainer.rs

```diff
@@ -0,0 +1,34 @@
+use crate::{trie::Nibbles, Bytes};
+use std::collections::BTreeMap;
+
+/// Proof retainer is used to store proofs during merkle trie construction.
+/// It is intended to be used within the [`HashBuilder`](crate::trie::HashBuilder).
+#[derive(Debug)]
+pub struct ProofRetainer {
+    targets: Vec<Nibbles>,
+    proofs: BTreeMap<Nibbles, Bytes>,
+}
```

Comment:
mind adding docs on the fields? like that the bytes should be rlp proofs, that the targets are prefixes (or entire addresses), etc